linfa-reduction/src/fast_ica/mod.rs

```diff
@@ -0,0 +1,382 @@
+        let xmean = x.mean_axis(Axis(0)).unwrap();
+        let mut xcentered = x - &xmean.to_owned().insert_axis(Axis(0));
+
+        // We transpose the centered matrix
+        xcentered = xcentered.reversed_axes();
+
+        // We whiten the matrix to remove any potential correlation between
+        // the components
+        let k = match xcentered.svd(true, false).unwrap() {
+            (Some(u), s, _) => {
+                let s = s.mapv(|x| A::from(x).unwrap());
+                (u.slice(s![.., ..nsamples.min(nfeatures)]).to_owned() / s)
+                    .t()
+                    .slice(s![..ncomponents, ..])
+                    .to_owned()
+            }
+            _ => unreachable!(),
+        };
+        let mut xwhitened = k.dot(&xcentered);
+
+        // We multiply the matrix with root of the number of records
+        let nsamples_sqrt = A::from((nsamples as f64).sqrt()).unwrap();
+        xwhitened.mapv_inplace(|x| x * nsamples_sqrt);
+
+        // We initialize the de-mixing matrix with a uniform distributiona
+        let w: Array2<f64>;
+        if let Some(seed) = self.random_state {
+            let mut rng = Isaac64Rng::seed_from_u64(seed as u64);
+            w = Array::random_using((ncomponents, ncomponents), Uniform::new(0., 1.), &mut rng);
+        } else {
+            w = Array::random((ncomponents, ncomponents), Uniform::new(0., 1.));
+        }
+        let mut w = w.mapv(|x| A::from(x).unwrap());
+
+        // We find the optimized de-mixing matrix
+        w = self.ica_parallel(&xwhitened, &w)?;
+
+        // We whiten the de-mixing matrix
+        let components = w.dot(&k);
+
+        Ok(FittedFastIca {
+            mean: xmean,
+            components,
+        })
+    }
+
+    // Parallel FastICA, Optimization step
+    fn ica_parallel<A: Float>(&self, x: &Array2<A>, w: &Array2<A>) -> Result<Array2<A>, String> {
+        let mut w = Self::sym_decorrelation(&w);
+
+        let p = x.ncols() as f64;
+
+        for _ in 0..self.max_iter {
+            let (gwtx, g_wtx) = self.gfunc.exec(&w.dot(x))?;
+
+            let lhs = gwtx.dot(&x.t()).mapv(|x| x / A::from(p).unwrap());
+            let rhs = &w * &g_wtx.insert_axis(Axis(1));
+            let wnew = Self::sym_decorrelation(&(lhs - rhs));
+
+            let lim = *wnew
+                .dot(&w.t())
+                .diag()
+                .mapv(num_traits::Float::abs)
+                .mapv(|x| x - A::from(1.).unwrap())
+                .mapv(num_traits::Float::abs)
+                .max()
+                .unwrap();
+
+            w = wnew;
+
+            if lim < A::from(self.tol).unwrap() {
+                break;
+            }
+        }
+
+        Ok(w)
+    }
+
+    // Symmetric decorrelation
+    //
+    // W <- (W * W.T)^{-1/2} * W
+    fn sym_decorrelation<A: Float>(w: &Array2<A>) -> Array2<A> {
+        let (eig_val, eig_vec) = w.dot(&w.t()).eigh(UPLO::Upper).unwrap();
```

Comment:
the eigenvalue decomposition could theoretical fail

Reply:
Now propagating the error instead of unwrapping